### PR TITLE
[RFC] V1 unit test rewrite top-down

### PR DESCRIPTION
**Pull request by JingwenGu0829**
Reorganizes the new unit-test contracts into component-focused folders so each file maps to the behavior it protects. This keeps the tests easier to review and extend without changing the runtime code paths under test.

Tried to make the contract simple and only protect the most important protocols for now.

## Test Folder Structure

```text
tests/unit_test/
  pipeline/
    helpers.py
    test_compile.py
    test_coordinator.py
    test_scheduler.py
    test_stage.py
  qwen3_omni/
    test_pipeline.py
    test_talker.py
    test_code2wav.py
  fishaudio_s2_pro/
    test_pipeline.py
    test_tts.py
    test_vocoder.py
```

#### pipeline

- `pipeline/test_compile.py`
  - Tests pipeline schema validation.
  - Tests config-to-runtime wiring, including runtime overrides, fan-in setup, and stream routing setup.
  - Tests MP runner TP rank setup and CUDA-visible-device process specs.

- `pipeline/test_coordinator.py`
  - Tests multi-terminal completion.
  - Tests abort propagation.
  - Tests fail-fast terminal errors and coordinator request-state cleanup.

- `pipeline/test_stage.py`
  - Tests per-request aggregation isolation.
  - Tests result routing and stream forwarding.
  - Tests relay tensor round-trips and cross-GPU stream metadata ordering.
  - Tests abort cleanup and relay-read failure reporting.

- `pipeline/test_scheduler.py`
  - Tests `SimpleScheduler` batch success output.
  - Tests per-request error emission when batch output contracts are violated.

#### Fish

- `fishaudio_s2_pro/test_pipeline.py`
  - Tests S2-Pro topology.
  - Tests state tensor serialization.
  - Tests tokenizer prompt/VQ layout.
  - Tests TTS request tensor fields and TTS result output-code shape.

- `fishaudio_s2_pro/test_tts.py`
  - Tests Fish model-runner VQ prompt embedding injection.
  - Tests semantic-code collection.
  - Tests scheduler chunked iteration, finished-result emission, and abort cleanup.

- `fishaudio_s2_pro/test_vocoder.py`
  - Tests vocoder batching.
  - Tests per-request audio trimming based on generated code length.

#### Qwen-Omni

- `qwen3_omni/test_pipeline.py`
  - Tests Qwen text/speech topology.
  - Tests `PipelineState` coercion.
  - Tests lightweight multimodal aggregation payloads and encoder-output pruning.
  - Tests thinker decode events.
  - Tests media-token hashing without changing M-RoPE input ids.

- `qwen3_omni/test_talker.py`
  - Tests talker feedback FIFO behavior.
  - Tests stream-done initialization.
  - Tests multimodal embedding injection and code-predictor output shape.

- `qwen3_omni/test_code2wav.py`
  - Tests incremental waveform emission.
  - Tests Code2Wav request-state cleanup on abort.

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
good!

File: tests/unit_test/fishaudio_s2_pro/test_vocoder.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import pytest
+import torch
+
+from sglang_omni_v1.models.fishaudio_s2_pro import stages
+from sglang_omni_v1.models.fishaudio_s2_pro.payload_types import S2ProState
+from sglang_omni_v1.scheduling.messages import IncomingMessage
+from tests.unit_test.fixtures.fish_fakes import FakeFishCodec, make_s2pro_payload
+
+
+def run_scheduler(
```
Comment (Ratish1) [inline tests/unit_test/fishaudio_s2_pro/test_vocoder.py]: We should reuse the existing scheduler helper instead of defining a second local `run_scheduler()` in `tests/unit_test/fishaudio_s2_pro/test_vocoder.py`. The helper in `tests/unit_test/pipeline/helpers.py` already starts a `SimpleScheduler`, pushes `IncomingMessage`s, drains a fixed output count, stops the scheduler, and joins the thread. The Fish vocoder file repeats the same lifecycle logic.

File: tests/unit_test/qwen3_omni/test_talker.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    scheduler = object.__new__(QwenTalkerScheduler)
+    req_data = SimpleNamespace(
+        pending_text_queue=deque([torch.tensor([11.0, 12.0])]),
+        thinker_chunks_done=True,
+    )
+    payload = SimpleNamespace(
+        prefetched_chunks=[SimpleNamespace(data=torch.tensor([20.0, 20.0]))],
+        prefetched_stream_done=True,
+    )
+    assert scheduler._is_request_build_ready(payload, pending_stream_done=True)
```
Comment (Ratish1) [inline tests/unit_test/qwen3_omni/test_talker.py:132]: Can we strengthen this test to cover the full talker streaming contract?

Right now it checks the happy path where feedback combines with the next text row, but it does not verify the boundary conditions that make the scheduler safe:

- `_is_request_build_ready(..., pending_stream_done=False)` should stay false, so the talker cannot build before the thinker stream is complete.
- `_take_next_decode_input_embed()` should pop both feedback and text after consuming them.
- feedback should remain queued when neither text nor final pad is available.
- the `tts_pad_embed` fallback should only be used after `thinker_chunks_done=True`.

**Comment by zhaochenyang20** [comment]
Under the new assumption that `V0` will be deleted, the old-stack IPC path can be ignored, but this file protects the live `sglang_omni_v1` IPC path. The trigger condition is a user starting a `V1` pipeline with `EndpointsConfig(scheme="ipc")`, where `create_ipc_runtime_dir`, `compile_pipeline_core`, `prepare_pipeline_runtime`, and `MultiProcessPipelineRunner` together allocate socket namespaces and clean them on failure or stop. The new `tests/unit_test/pipeline/test_compile.py` only uses `EndpointsConfig(scheme="tcp")`; it covers compile wiring, TP ranks, and CUDA-visible-device process specs, but it does not cover IPC namespace uniqueness, owned runtime-dir cleanup on failure, caller-owned runtime dirs not being deleted, MP runner startup-failure cleanup, or endpoint non-reuse across two same-name pipeline instances.

If the `V1` IPC branch regresses later, the new fast lane will not fail. The concrete failure modes are socket directories leaking after startup failures, same-name pipelines reusing endpoints, or caller-owned runtime-dir ownership being reversed. These usually surface only under multiprocess serving or repeated starts of pipelines with the same model name.

The old file location does not need to remain, but its non-model `V1` IPC lifecycle cases should be ported into `tests/unit_test/pipeline/test_ipc_runtime_dir.py` or `test_compile.py`. The tests can still use fake factories, `tempfile.TemporaryDirectory`, and patched coordinators; the important contracts are IPC dir uniqueness, compile-failure cleanup, caller-owned-dir preservation, MP runner startup-failure cleanup, stop cleanup, and same-model endpoint uniqueness.

**Comment by zhaochenyang20** [comment]
https://github.com/sgl-project/sglang-omni/pull/380

will be moved here

File: tests/test_v1_threaded_simple_scheduler.py
Comment (zhaochenyang20) [inline tests/test_v1_threaded_simple_scheduler.py:1]: This deletion is still a `V1` risk. The runtime path is direct: FishAudio S2-Pro `create_preprocessing_executor` still returns `ThreadedSimpleScheduler(_preprocess, max_concurrency=max_concurrency)` for blocking CPU preprocessing. The new `tests/unit_test/pipeline/test_scheduler.py` only covers `SimpleScheduler` batch success and batch failure; it does not exercise the thread pool, capacity gating, future callbacks, executor shutdown, or whether compute exceptions raised in worker threads are emitted into the outbox.

If the scheduler regresses to serial execution, Fish preprocessing throughput drops; if exceptions are only logged and not emitted as `OutgoingMessage(type="error")`, upstream stages and the coordinator will not observe terminal errors. The current `tests/unit_test` suite can pass while both regressions exist.

Move the old concurrency and error cases into `tests/unit_test/pipeline/test_scheduler.py`, or add `tests/unit_test/pipeline/test_threaded_scheduler.py`. The concurrency case can keep using two blocking requests, `threading.Event`, and `max_concurrency=2`; the error case should assert `request_id`, `type == "error"`, and a `RuntimeError` instance.

File: tests/unit_test/qwen3_omni/test_pipeline.py
Comment (zhaochenyang20) [inline tests/unit_test/qwen3_omni/test_pipeline.py:1]: The new test compresses multiple Qwen talker states into one positive assertion, only checking `_is_request_build_ready(..., pending_stream_done=True)`. The real trigger path is that the talker scheduler must wait for the thinker stream to finish before building a request; during decode, a batch may continue only when the feedback queue has an element and either a text chunk is available or thinker completion allows a pad embedding. The old `test_v1_talker_feedback.py` covered the `pending_stream_done=False` false case, pad fallback after stream completion, preserving feedback when text is absent, and the no-text/with-text/with-pad states for `_data_has_next_decode_input`. The new test keeps only the no-text `None` result and later text-arrival happy path, so it no longer locks down the full build-readiness and decode-readiness state space.

An unconditional true return from `_is_request_build_ready` would let the talker consume an incomplete thinker stream; an incorrect `_take_next_decode_input_embed` or `_data_has_next_decode_input` could pop feedback too early, skip pad fallback, or release decode batches incorrectly. The result can be speech tokens misaligned with text context, premature EOS, or a corrupted feedback FIFO.

Restore these states inside the new `tests/unit_test/qwen3_omni/test_talker.py`; the old filename does not need to come back. At minimum, add back the `pending_stream_done=False` assertion, the stream-done `tts_pad_embed` fallback case, the assertion that feedback is not popped when text is absent and thinker is not done, and `_data_has_next_decode_input` checks for no-text, with-text, and with-pad states.

```python
assert not scheduler._is_request_build_ready(payload, pending_stream_done=False)
assert scheduler._is_request_build_ready(payload, pending_stream_done=True)
```

File: tests/test_v1_fish_s2pro_runtime.py
Comment (zhaochenyang20) [inline tests/test_v1_fish_s2pro_runtime.py:1]: This deleted file protects the current `V1` Fish scheduler/model-runner state machine, not `V0`. The real runtime path is `FishS2ProModelRunner.collect_s2pro_step_outputs` collecting semantic tokens and audio codebook rows, `FishIterationController` deciding completion from `im_end_token_id`, chunked state, and output-code length, and `FishScheduler.emit_finished` emitting an error before the result adapter/vocoder when no audio codec tokens were generated. The new `tests/unit_test/fishaudio_s2_pro/test_tts.py` covers VQ injection, ordinary code collection, chunked decrement, finished result emission, and abort cleanup, but it does not cover immediate `im_end` producing no audio frame, terminal and audio requests staying separated in the same batch, or empty `output_codes` producing an error before the vocoder path.

If terminal EOS is treated as an audio timestep, or a terminal request in a mixed batch contaminates an audio request's `last_codebook_values` / `previous_semantic_tokens`, S2-Pro can send the wrong codebook frame to the vocoder. If immediate-EOS empty `output_codes` is no longer converted to an error first, the non-empty assertion in `apply_tts_result` will fail deeper in the stack, producing a later and harder-to-diagnose error.

Move the old cases that still protect `V1` production contracts into `tests/unit_test/fishaudio_s2_pro/test_tts.py`: terminal `im_end` does not append an audio codebook frame, immediate `im_end` leaves `output_codes` empty, empty `output_codes` emits a `ValueError` in `emit_finished` without calling the result adapter, terminal/audio request state stays separated in mixed batches, and chunked steps do not mutate decode state. The existing `_FakePlanner`, `_FakeResourceManager`, and `FakeFishReq` can be reused; the old test structure does not need to return.

File: tests/unit_test/fishaudio_s2_pro/test_vocoder.py
Comment (zhaochenyang20) [inline tests/unit_test/fishaudio_s2_pro/test_vocoder.py:1]: This local `run_scheduler` duplicates the shared helper at `tests/unit_test/pipeline/helpers.py:51`. Both trigger paths start a blocking scheduler loop, enqueue inbox messages, read a fixed number of outbox messages, and stop/join in `finally`. Since this PR already extracted scheduler lifecycle into a shared helper, keeping a Fish vocoder copy creates two maintenance points for timeout, stop/join, and any future draining behavior.

This is not a correctness blocker, but it lets the same scheduler-test pattern drift. If the shared helper later fixes a thread leak or changes timeout policy, the Fish vocoder test will not benefit and may behave differently from the pipeline scheduler tests in CI.

 Remove the local `run_scheduler` and the `threading` import, then import `run_scheduler` from `tests.unit_test.pipeline.helpers`. The current call already uses `output_count=2`, which is compatible with the shared helper's keyword-only argument.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Comment by zhaochenyang20** [comment]
Since we will directly remove V0 this week, I think we should not care about protecting any feature of V0. We can just focus on removing them, but we need to care about every feature of V1.

File: tests/unit_test/qwen3_omni/test_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+            "thinker_out": {"output_ids": [3], "is_final": True},
+        }
+    )
+    assert torch.equal(state.prompt["input_ids"], torch.tensor([1, 2]))
+    assert state.mm_inputs == {}
+    assert state.encoder_inputs["image_encoder"]["cache_key"] == "img"
+    assert state.thinker_out["is_final"] is True
+
+
+def test_qwen_cli_mem_fraction_precedence_targets_only_ar_stages() -> None:
```
Comment (zhaochenyang20) [inline tests/unit_test/qwen3_omni/test_pipeline.py]: The new `test_qwen_cli_mem_fraction_precedence_targets_only_ar_stages` covers only one path: global `--mem-fraction-static` falls back to the talker, `--thinker-mem-fraction-static` overrides the thinker, and non-AR stages are left untouched. The trigger condition is a user launching a Qwen speech pipeline through `sglang_omni_v1.cli.serve`, where `apply_mem_fraction_cli_overrides`, `apply_encoder_mem_reserve_cli_override`, `build_sglang_server_args`, and the Qwen thinker/talker factories together determine SGLang `ServerArgs.mem_fraction_static`. After deleting the old `tests/test_v1_mem_fraction_static.py`, the new tests no longer provide equivalent coverage for `--talker-mem-fraction-static` overriding the global flag, text-only Qwen rejecting talker flags without partial writes, invalid fractions failing before config mutation, global flags failing on pipelines with no supported role, builder default omission of `mem_fraction_static`, explicit builder forwarding of `mem_fraction_static`, valid and invalid `apply_encoder_mem_reserve` boundaries, `--encoder-mem-reserve` exclusivity with global/thinker pins, config/runtime pinned thinker memory rejection, or reserve remaining thinker-only after runtime override overlay.

The old filename does not need to return, but the public-boundary cases from the old file should be ported into `tests/unit_test/qwen3_omni/test_pipeline.py`, or split into `tests/unit_test/qwen3_omni/test_memory_flags.py`. At minimum, restore builder default and explicit forwarding, the full global/per-role precedence combinations, talker override, invalid fraction/no-partial-write behavior, unsupported-role rejection, encoder-reserve exclusivity, config/runtime pinned thinker rejection, runtime overlay behavior, and the safe-floor case for `_apply_qwen_thinker_encoder_reserve`. That keeps the test reorganization from downgrading user-visible launch-parameter contracts to one happy path.

File: tests/unit_test/pipeline/test_ipc.py
Comment (zhaochenyang20) [inline tests/unit_test/pipeline/test_ipc.py:1]: The new IPC test imports the same modules both as module aliases and through direct symbols: `import sglang_omni_v1.config.compiler as compiler` plus `from sglang_omni_v1.config.compiler import ...`, and `import sglang_omni_v1.pipeline.mp_runner as mp_runner` plus `from sglang_omni_v1.pipeline.mp_runner import _build_stage_groups`. The trigger path is that later tests both monkeypatch the module object and call directly imported symbols. This does not change runtime behavior, but it splits patch targets and call sites into two names; if compiler or mp_runner later renames a symbol, wraps it, or needs `_build_stage_groups` patched, readers must verify whether the direct import captured an old reference, which creates avoidable maintenance risk.

Keep the module-alias imports because these tests already need `monkeypatch.setattr(compiler, ...)` and `monkeypatch.setattr(mp_runner, ...)`. Remove the direct imports, then qualify calls to `create_ipc_runtime_dir`, `compile_pipeline`, `compile_pipeline_core`, `prepare_pipeline_runtime`, and `_build_stage_groups` with `compiler.` or `mp_runner.`. That keeps patch targets, call paths, and import style consistent.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: tests/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+├── docs/
+│   ├── qwen3_omni/
+│   └── s2pro/
+├── test_model/
+└── unit_test/
+    ├── fixtures/
+    │   ├── fish_fakes.py
+    │   ├── pipeline_fakes.py
+    │   └── qwen_fakes.py
+    ├── pipeline/
```
Comment (zhaochenyang20) [inline tests/README.md:17]: `tests/README.md` is the entry document this test-reorganization PR gives future contributors, but the current tree under `unit_test/pipeline/` lists only `helpers.py`, `test_compile.py`, `test_coordinator.py`, `test_scheduler.py`, and `test_stage.py`. The verified head now also has `tests/unit_test/pipeline/test_ipc_runtime_dir.py` and `tests/unit_test/pipeline/test_threaded_scheduler.py`, while the ownership text does not include IPC runtime-dir lifecycle, MP runner cleanup, or `ThreadedSimpleScheduler` concurrency/error contracts as pipeline responsibilities. The trigger condition is practical: the next contributor choosing where to put a new `V1` test from the README sees a stale architecture diagram and an overly broad ownership description. The impact is that new tests can keep accumulating in oversized `test_pipeline.py` files, or public CLI/config/builder contracts can land in the wrong lane, leaving future reviews to recover coverage by memory.

Update the README tree so it lists the files that actually exist, at least including `test_ipc_runtime_dir.py` and `test_threaded_scheduler.py`. More importantly, add a "how to add a test" rule set: model-agnostic compile/coordinator/stage/relay/IPC/scheduler tests go under `unit_test/pipeline/`; Qwen public CLI/config/builder, `PipelineState`, request builders, talker, and code2wav tests go under `unit_test/qwen3_omni/`, with a dedicated `test_memory_flags.py` if needed; S2-Pro tokenizer/state/TTS scheduler/vocoder tests go under `unit_test/fishaudio_s2_pro/`; real server/model/GPU tests go under `test_model/`; docs examples go under `docs/` and use the docs marker; shared fakes belong in `fixtures/` only when reused, and single-test helpers should not be prematurely promoted. That makes the README guide contributors toward tests that match the current architecture instead of merely documenting a static tree.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: tests/unit_test/qwen3_omni/test_pipeline.py [hunk 5]
```diff
@@ -343,3 +410,131 @@ def test_qwen_factory_signatures_keep_reserve_thinker_only() -> None:
...
+        }
+    )
+    merged_state = PipelineState.from_dict(merged.data)
+    assert merged_state.encoder_inputs == {}
+    assert merged_state.encoder_outs == {}
+    assert "image_embeds" in merged_state.thinker_inputs["model_inputs"]
+    assert merged_state.thinker_inputs["media_cache_keys"] == {
+        "image": "image:image-cache",
+        "video": "video:image-cache",
+    }
```
Comment (zhaochenyang20) [inline tests/unit_test/qwen3_omni/test_pipeline.py:456]: The trigger condition is a user request with audio input: `project_preprocessing_to_audio_encoder` creates `encoder_inputs["audio_encoder"]["cache_key"]`, then `merge_for_thinker`/`build_thinker_inputs` must convert it into `media_cache_keys["audio"]`, and `build_sglang_thinker_request` uses that key to hash the audio placeholder token. The new `test_qwen_mm_aggregate_keeps_lightweight_inputs_and_prunes_after_merge` only constructs an `image_encoder` cache key and asserts the `image` and `video` keys; the deleted `tests/test_v1_qwen3_omni_thinker_prefix_cache.py` covered `audio_encoder` -> `audio:aud-cache`, but that branch has no equivalent in the new layout. The impact is that a future regression in audio cache-key construction would still pass the new `tests/unit_test` lane, allowing Qwen audio prompts to use an unhashed or wrongly hashed placeholder and drift away from the prefix-cache semantics.

Add an audio encoder cache-key input to the existing Qwen pipeline test, or add a focused `build_thinker_inputs` case. The minimal fix is to include `"audio_encoder": {"cache_key": "audio-cache"}` in `state.encoder_inputs` and assert that the merged state's `media_cache_keys` contains `"audio": "audio:audio-cache"`. That reuses the current `merge_for_thinker` path and does not require restoring the old filename.

File: tests/unit_test/qwen3_omni/test_talker.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+
+def test_qwen_model_runner_and_code_predictor_tensor_contracts() -> None:
+    """Preserves multimodal embed injection and code-predictor token shape."""
+
+    class RecordingEmbed:
+        num_embeddings = 10
+
+        def __init__(self) -> None:
+            self.seen: torch.Tensor | None = None
```
Comment (zhaochenyang20) [inline tests/unit_test/qwen3_omni/test_talker.py:225]: The trigger condition is the Qwen speech pipeline building talker prefill early in the thinker stream, receiving a late text chunk after the thinker has already finished, or feeding HF-style 4D logits into the residual code predictor. The current `test_qwen_talker_decode_*` cases restore feedback FIFO and decode gating, but `test_qwen_model_runner_and_code_predictor_tensor_contracts` only keeps the 3D-logits argmax shape; the short-prefix placeholder layout from `build_assistant_part`, `TalkerPrefillBuilder.append_text_chunk` ignoring late chunks when `thinker_chunks_done=True`, and the 4D-logits no-extra-unsqueeze case from the deleted V1 tests are not migrated. The impact is that these live V1 talker boundaries can regress without failing the fast unit lane: short prefixes can break the 9-row assistant layout, late chunks can pollute a completed pending text queue, and a 4D shape regression can add an extra code-predictor dimension and break downstream decode.

Restore the old cases as componentized tests in `tests/unit_test/qwen3_omni/test_talker.py`. Reuse `tests.unit_test.fixtures.qwen_fakes.FakeTalkerProjectionModel` or small local fakes to cover: `build_assistant_part` with a 3-token assistant prefix, including the zero fourth-token slot and `future_text_rows`; `TalkerPrefillBuilder.append_text_chunk` not appending when `thinker_chunks_done=True`; and `Qwen3OmniTalker._sample_code_predictor_token` returning `(1, 2)` rather than `(1, 2, 1)` for 4D logits.

File: tests/unit_test/pipeline/test_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    """Preserves batched success output and per-request batch failure emission."""
+    good = SimpleScheduler(
+        lambda payload: payload,
+        batch_compute_fn=lambda payloads: [payload.upper() for payload in payloads],
+        max_batch_size=2,
+        max_batch_wait_ms=10,
+    )
+    outputs = run_scheduler(
+        good,
+        [
```
Comment (zhaochenyang20) [inline tests/unit_test/pipeline/test_scheduler.py:25]: The trigger condition is a V1 AR stage using the generic `OmniScheduler` without a custom `stream_chunk_handler` or `stream_done_handler`. In that path `_append_stream_chunk_default` should buffer the raw chunk in `req_data.stream_chunks`, and `_mark_stream_done` should set the generic `req_data.stream_done=True`; generic AR scheduler paths such as Qwen thinker and Fish TTS rely on that default semantic staying stable. The new scheduler suite only covers `SimpleScheduler` batch/error and `ThreadedSimpleScheduler` concurrency/error, while the two default stream cases from the deleted `tests/test_v1_omni_scheduler.py` were not moved into `unit_test/pipeline/`. The impact is that a regression in the generic scheduler stream fallback would pass the new layout, and a stage without a custom handler could stop seeing upstream stream chunks or stream completion correctly.

Add two small `OmniScheduler` tests to `tests/unit_test/pipeline/test_scheduler.py`: call `OmniScheduler._append_stream_chunk_default(SimpleNamespace(), chunk)` directly and assert the raw chunk is preserved; then create `object.__new__(OmniScheduler)`, set `_stream_done_handler=None`, call `_mark_stream_done`, and assert `stream_done is True`. These tests belong to the README's model-agnostic scheduler behavior lane and do not need a new helper.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Review by zhaochenyang20 — APPROVED** [review]
